module/zfs/vdev_raidz.c

```diff
@@ -2234,6 +2234,33 @@ vdev_raidz_get_logical_width(vdev_raidz_t *vdrz, uint64_t txg)
 	mutex_exit(&vdrz->vd_expand_lock);
 	return (width);
 }
+/*
+ * This code converts an asize into the largest psize that can safely be written
+ * to an allocation of that size for this vdev.
+ * 
+ * Note that this function will not take into account the effect of gang
+ * headers, which also modify the ASIZE of the DVAs. It is purely a reverse of
+ * the psize_to_asize function.
+ */
+static uint64_t
+vdev_raidz_asize_to_psize(vdev_t *vd, uint64_t asize, uint64_t txg)
+{
+	vdev_raidz_t *vdrz = vd->vdev_tsd;
+	uint64_t psize;
+	uint64_t ashift = vd->vdev_top->vdev_ashift;
+	uint64_t cols = vdrz->vd_original_width;
+	uint64_t nparity = vdrz->vd_nparity;
+
+	cols = vdev_raidz_get_logical_width(vdrz, txg);
+
+	ASSERT0(asize % (1 << ashift));
+
+	psize = (asize >> ashift);
+	psize -= nparity * ((asize + cols - nparity - 1) / (cols - nparity));
```

Comment:
```suggestion
	psize -= nparity * DIV_ROUND_UP(psize, cols - nparity);
```

Reply:
So the psize-to-asize steps here are:
1. divide by 1<<ashift, rounded up  (How many sectors do we need?)
2. add nparity * CEIL(asize / (cols - nparity)) (How many rows do that many sectors take up? Add the parity for that many rows)
3. round up to a multiple of nparity + 1 (RAIDZ roundup)
4. Shift left by ashift (back to bytes)

So the inverse steps should be
1. Shift right by ashift (convert to sectors)
2. Nothing (We can't un-round-up to nparity + 1)
3. Subtract nparity * CEIL(asize/cols) (How many rows were we taking up? Subtract out the parity for those rows)
4. Shift right by ashift (convert back to bytes)

All of which is to say: Yes, you're right, but I wanted to write it all out so I could actually make sure I understood the math.